dbms/src/DataStreams/HashJoinProbeBlockInputStream.cpp

```diff
@@ -67,17 +68,81 @@ Block HashJoinProbeBlockInputStream::getHeader() const
 
 Block HashJoinProbeBlockInputStream::readImpl()
 {
-    if (probe_process_info.all_rows_joined_finish)
+    // if join finished, return {}
+    if (join_finished)
+    {
+        return Block{};
+    }
+
+    result_blocks.clear();
+    size_t output_rows = 0;
+
+    // if over_limit_block is not null, we need to push it into result_blocks first.
+    if (over_limit_block)
```

Comment:
Can we use [SquashingTransform](https://github.com/pingcap/tiflash/blob/master/dbms/src/DataStreams/SquashingTransform.h) directly?

Reply:
If not, can it be a separate class?

Reply:
> If not, can it be a separate class?

done.